lib/internal/quic/binding.js

```diff
@@ -0,0 +1,209 @@
+'use strict';
+
+const {
+  initializeCallbacks,
+} = internalBinding('quic');
+
+// If the initializeCallbacks is undefined, the Node.js binary
+// was built without QUIC support, in which case we
+// don't want to export anything here.
+if (initializeCallbacks === undefined)
+  return;
+
+const {
+  symbols: {
+    owner_symbol,
+  },
+} = require('internal/async_hooks');
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_STATE,
+    ERR_QUIC_APPLICATION_ERROR,
+    ERR_QUIC_VERSION_NEGOTIATION,
+  },
+} = require('internal/errors');
+
+const {
+  isAnyArrayBuffer,
+  isArrayBufferView,
+} = require('internal/util/types');
+
+const {
+  kAddSession,
+  kClose,
+  kCreatedStream,
+  kHandshakeComplete,
+  BlockedEvent,
+  ClientHelloEvent,
+  KeylogEvent,
+  OCSPRequestEvent,
+  OCSPResponseEvent,
+  SessionTicketEvent,
+} = require('internal/quic/common');
+
+const assert = require('internal/assert');
+
+function onEndpointError(error) {
+  assert(error !== undefined);
+  this[owner_symbol].destroy(error);
+}
+
+function onEndpointDone() {
+  this[owner_symbol].destroy();
+}
+
+function onSessionNew(sessionHandle) {
+  this[owner_symbol][kAddSession](sessionHandle);
+}
+
+function onSessionClientHello(alpn, servername, ciphers) {
+  let alreadyDone = false;
+  this[owner_symbol].dispatchEvent(
+    new ClientHelloEvent(alpn, servername, ciphers, (context) => {
+      if (context !== undefined &&
+          typeof context?.context?.external !== 'object') {
```

Comment:
Duplicate `.context?`? (just to make sure! 😄)

Reply:
Yes, this is correct. It's checking for `SecureContext` and this is the only way to do it reliably right now.